File: medium/016_3Sum_Closest/3sum_closest.py
class Solution(object):
    def add_to_result(self, result, total, target):
        if len(result) == 0:
            result.append(total)
        elif abs(total-target) < abs(result[0]-target):
            result[0] = total

    def two_sum_closest(self, a, start, x0, target, result):
        i, j = start, len(a)-1
        while i < j:
            total = x0 + a[i] + a[j]
            self.add_to_result(result, total, target)
            if total == target:
                return
            elif total > target:
                j -= 1
            else:
                i += 1

    def threeSumClosest(self, a, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        result = []
        a.sort()
        for i in xrange(len(a)):
            self.two_sum_closest(a, i+1, a[i], target, result)
        return result[0]

def test_three_sum_closest(a, target):
    result = solution.threeSumClosest(a, target)
    print "In {}, 3 sum closest to {} is {}".format(a, target, result)

solution = Solution()
test_three_sum_closest([-1, 2, 1, -4], 1)

File: medium/016_3Sum_Closest/3sumClosest.go
package main

import "fmt"
import "sort"

func abs(a int) int {
	if a >= 0 {
		return a
	} else {
		return -a
	}
}

func twoSumClosest(nums []int, target int) int {
	i, j := 0, len(nums)-1
	closeSum := nums[i] + nums[j]
	for i < j {
		sum := nums[i] + nums[j]
		if abs(target-sum) < abs(target-closeSum) {
			closeSum = sum
		}
		if sum > target {
			j--
		} else if sum < target {
			i++
		} else {
			break
		}
	}
	return closeSum
}

func threeSumClosest(nums []int, target int) int {
	var closeSum int
	sort.Ints(nums)
	for i := 0; i < len(nums)-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		res := twoSumClosest(nums[i+1:], target-nums[i])
		if i == 0 {
			closeSum = nums[i] + res
		} else {
			sum := nums[i] + res
			if abs(target-sum) < abs(target-closeSum) {
				closeSum = sum
			}
		}
		if closeSum == target {
			break
		}
	}
	return closeSum
}

func testThreeSumClosest(nums []int, target int) {
	res := threeSumClosest(nums, target)
	fmt.Printf("target %d, %v, get %v\n", target, nums, res)
}

func main() {
	testThreeSumClosest([]int{}, 0)
	testThreeSumClosest([]int{1}, 0)
	testThreeSumClosest([]int{-1, 2, 1, -4}, 1)
}
